linux_os/guide/services/ssh/ssh_server/sshd_do_not_permit_user_env/oval/shared.xml

```diff
@@ -35,6 +46,6 @@
   <ind:textfilecontent54_object id="obj_sshd_no_user_envset" version="2">
     <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
     <ind:pattern operation="pattern match">^[\s]*(?i)PermitUserEnvironment(?-i)[\s]+no[\s]*(?:|(?:#.*))?$</ind:pattern>
-    <ind:instance datatype="int">1</ind:instance>
+    <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
```

Comment:
Here it is the same as `StrictModes yes`. We don't need to have the specific operation here, but instead change the variable in line 41 to `check_existence="all_exist"`.

Reply:
Hmm... I'm confused.

My understanding was that `operation="greater than or equal"` was required to consider more than the first matching line, but from my testing now that doesn't seem to be the case. My commit does indeed seem quite pointless than. At least I can't construct any case now where it'd make a difference.

Ccurrently, the test has `none_exist`, but it's referenced with `negate="true"`. That should be equivalent to `all` without the negation.

Anyway, I think what we'd really want to do here is:
* find all lines with `PermitUserEnvironment`
* use `all_exist` to make sure that there's at least one such line
* use a state to make sure they are all set to `no` (by matching a `subexpression`)

That makes sure there is no `PermitUserEnvironment yes` anywhere later in the file.

Reply:
> Hmm... I'm confused.
> 
> My understanding was that `operation="greater than or equal"` was required to consider more than the first matching line, but from my testing now that doesn't seem to be the case. My commit does indeed seem quite pointless than. At least I can't construct any case now where it'd make a difference.

From my understanding, the value `1` here means that the first occurrence that matches the regular expression will be fetched. If you put `2`, then the second will be fetched and so on. So, `greater than or equal` kinda doesn't make sense here. Or I am missing something related to OVAL.
> 
> Ccurrently, the test has `none_exist`, but it's referenced with `negate="true"`. That should be equivalent to `all` without the negation.

Agree, I just feel that `all_exist` is easier to understand, but whatever solution we go for, it would be nice to change in all other rules which touch sshd configuration. Currently I opened #4424 to fix rules which touch sshd configuration and weren't according to STIG recommendation.
> 
> Anyway, I think what we'd really want to do here is:
> 
>     * find all lines with `PermitUserEnvironment`
> 
>     * use `all_exist` to make sure that there's at least one such line
> 
>     * use a state to make sure they are all set to `no` (by matching a `subexpression`)
> 
> 
> That makes sure there is no `PermitUserEnvironment yes` anywhere later in the file.

That would be the best approach. I got a suggestion in the aforementioned PR (#4424) on how to handle this situation. I'll try out and see if it works.

Reply:
I've just backed out my (wrong) change. If you're changing this file in #4424, then it seems best to me to do it all together.

Reply:
Ok. We can work on this. But first there is a discussion ongoing on #4424 whether we will check for presence of right configuration (`StrictModes yes`) or absence of wrong configuration (`StrictModes no`). We are probably going for only checking for undesired configuration.

Reply:
All right, based on the conversation in #4424 we can proceed by checking the absence of non-default configuration. This imply that, for example `StrictModes no` should not be present in the configuration. So we'll be using `none_exist` and regular expression for `StrictModes no`.

I've seen that for `sle` you previously wanted to check differently, like checking the presence of right configuration, if this is a hard requirement feel free to object and disagree.